### DISCUSSION TITLE
LSP Question when using Xcode generators

### DISCUSSION THREAD

**Discussion by lucianthorr** [microsoft/vscode-cmake-tools#4222] [General]
I'm really not sure if this is a question for the cmake-tools extension or the cpptools extension so forgive me if I'm in the wrong place.

Can anyone explain or point me in a direction for how the VSCode LSP features (intellisense, etc) work for CMake C++ projects that are using Xcode as the generator?  From my understanding the clangd LSP relies on a `compile_commands.json` file for project file information but from my understanding Xcode doesn't generate one.  How are the VSCode extensions able to work without that piece of the puzzle?
Thanks.

**Comment by gcampbell-msft** [comment]
@Colengms @sean-mcmanus Looping folks in that can help provide context to this conversation. They are maintainers of the C++ extension.

**Reply by sean-mcmanus** [reply]
The C/C++ extension doesn't interact with CMake generators. It just handles the vscode-cpptools-api that the CMake Tools extension provides and the CMake Tools extension handles the XCode generator (I'm not familiar with how it does it, but you could possibly look into the OSS code in this repo). When the configurationProvider setting it set then compile_commands.json are not processed by the C/C++ extension if the configuration provider has a configuration for a file.

**Comment by bobbrow** [comment]
@lucianthorr there are no LSP features tied specifically to a C++ configuration so every extension out there has its own way of doing things.

`clangd` requires a `compile_commands.json` file as you mentioned, but when you use the XCode generator CMake doesn't create one.  If you are able to configure your project with the Ninja or Makefile generators, you can get CMake to write a `compile_commands.json` file for you by using the [CMAKE_EXPORT_COMPILE_COMMANDS](https://cmake.org/cmake/help/latest/variable/CMAKE_EXPORT_COMPILE_COMMANDS.html) variable.

Microsoft's C++ extension communicates directly with the CMake Tools extension as @sean-mcmanus mentioned so it gets the information from CMake's File API instead of `compile_commands.json`.  CMake is capable of providing configurations for all generators in this way, so the Microsoft C++ extension is able to remain independent of the configuration process.

**Reply by lucianthorr** [reply]
@bobbrow @sean-mcmanus Thanks for taking the time.  Both extensions are great also so thanks for all the work.
I may follow up with another question or two but I'd do some digging through the code first and see what I can figure out.
